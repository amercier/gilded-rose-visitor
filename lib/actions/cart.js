import { CART_ADD_ITEM, CART_REMOVE_ITEM } from '../constants/actionTypes';

/**
 * Create an action to add an item in the cart.
 *
 * @param {string} id - Item id.
 * @returns {Object} An action that consists in adding an item in the cart..
 */
export const doAddItemToCart = id => ({
  type: CART_ADD_ITEM,
  id,
});

/**
 * Create an action to remove an item from the cart.
 *
 * @param {string} id - Item id.
 * @returns {Object} An action that consists in removing an item from the cart.
 */
export const doRemoveItemFromCart = id => ({
  type: CART_REMOVE_ITEM,
  id,
});
