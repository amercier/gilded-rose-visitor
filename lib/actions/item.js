import {
  ITEMS_POLL_START,
  ITEMS_POLL_STOP,
  ITEMS_FETCH,
  ITEMS_FETCH_ERROR,
  ITEMS_SET,
  ITEM_FETCH,
  ITEM_FETCH_ERROR,
  ITEM_SET,
} from '../constants/actionTypes';
import { ITEMS_POLL_DELAY } from '../constants/config';

/**
 * Create an action to start polling items every ITEMS_POLL_DELAY ms, until a ITEMS_POLL_STOP action
 * is triggered.
 *
 * @returns {Object} An action that consists in starting polling items.
 */
export const doStartPollingItems = () => ({
  type: ITEMS_POLL_START,
  delay: ITEMS_POLL_DELAY,
});

/**
 * Create an action to stop polling items.
 *
 * @returns {Object} An action that consists in stopping polling items.
 */
export const doStopPollingItems = () => ({
  type: ITEMS_POLL_STOP,
});

/**
 * Create an action to fetch items.
 *
 * @returns {Object} An action that consists in fetching items.
 */
export const doFetchItems = () => ({
  type: ITEMS_FETCH,
});

/**
 * Create an action to handle an items fetch error.
 *
 * @param {Error} error - Error thrown during fetch.
 * @returns {Object} An action that consists in hnadling items fetch error.
 */
export const doFetchItemsError = error => ({
  type: ITEMS_FETCH_ERROR,
  error,
});

/**
 * Create an action to add items.
 *
 * @param {Item[]} items - Items to add.
 * @returns {Object} An action that consists in adding the given items.
 */
export const doSetItems = items => ({
  type: ITEMS_SET,
  items,
});

/**
 * Create an action to fetch an item's details.
 *
 * @param {string} id - The item id.
 * @returns {Object} An action that consists in fetching items.
 */
export const doFetchItem = id => ({
  type: ITEM_FETCH,
  id,
});

/**
 * Create an action to handle an items fetch error.
 *
 * @param {Error} error - Error thrown during fetch.
 * @returns {Object} An action that consists in hnadling items fetch error.
 */
export const doFetchItemError = error => ({
  type: ITEM_FETCH_ERROR,
  error,
});

/**
 * Create an action to add items.
 *
 * @param {Item} item - Items to add.
 * @returns {Object} An action that consists in adding the given item.
 */
export const doSetItem = item => ({
  type: ITEM_SET,
  item,
});
