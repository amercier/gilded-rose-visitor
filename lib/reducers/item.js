import {
  ITEMS_FETCH,
  ITEMS_FETCH_ERROR,
  ITEMS_SET,
  ITEM_FETCH,
  ITEM_FETCH_ERROR,
  ITEM_SET,
} from '../constants/actionTypes';

const INITIAL_STATE = {
  items: {},
  fetchingItems: false,
  fetchedItemsOnce: false,
  fetchItemsError: undefined,
  item: {},
  fetchingItem: false,
  fetchItemError: undefined,
};

/**
 * Item reducer.
 *
 * @param {Object} [state=INITIAL_STATE] - Previous state.
 * @param {string} [action={}] - Action type and data.
 * @returns {Object} Next state.
 */
function itemReducer(state = INITIAL_STATE, action = {}) {
  switch (action.type) {
    case ITEMS_FETCH: {
      return { ...state, fetchingItems: true };
    }
    case ITEMS_FETCH_ERROR: {
      // TODO Display a notification when the error occur while items exist.
      // TODO Detect whether Internet connection is lost (normal behavior) vs server error.
      return {
        ...state,
        fetchingItems: false,
        fetchedItemsOnce: true,
        fetchItemsError: action.error,
      };
    }
    case ITEMS_SET: {
      return {
        ...state,
        items: action.items,
        fetchingItems: false,
        fetchedItemsOnce: true,
        fetchItemsError: undefined,
      };
    }
    case ITEM_FETCH: {
      return { ...state, fetchingItem: true };
    }
    case ITEM_FETCH_ERROR: {
      return {
        ...state,
        fetchingItem: false,
        fetchItemError: action.error,
      };
    }
    case ITEM_SET: {
      return {
        ...state,
        item: action.item,
        fetchingItem: false,
        fetchItemError: undefined,
      };
    }
    default:
      return state;
  }
}

export default itemReducer;
