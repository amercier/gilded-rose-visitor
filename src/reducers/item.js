import {
  ITEMS_FETCH,
  ITEMS_FETCH_ERROR,
  ITEMS_SET,
} from '../constants/actionTypes';

const INITIAL_STATE = {
  items: {},
  fetchingItems: false,
  fetchedItemsOnce: false,
  fetchItemsError: undefined,
};

/**
 * Item reducer.
 *
 * @param {Object} [state=INITIAL_STATE] - Previous state.
 * @param {string} [action={}] - Action type and data.
 * @returns {Object} Next state.
 */
function itemReducer(state = INITIAL_STATE, action = {}) {
  switch (action.type) {
    case ITEMS_FETCH: {
      return { ...state, fetchingItems: true };
    }
    case ITEMS_FETCH_ERROR: {
      // TODO Display a notification when the error occur while items exist.
      // TODO Detect whether Internet connection is lost (normal behavior) vs server error.
      return {
        ...state,
        fetchingItems: false,
        fetchedItemsOnce: true,
        fetchItemsError: action.error,
      };
    }
    case ITEMS_SET: {
      return {
        ...state,
        items: action.items,
        fetchingItems: false,
        fetchedItemsOnce: true,
        fetchItemsError: undefined,
      };
    }
    default:
      return state;
  }
}

export default itemReducer;
