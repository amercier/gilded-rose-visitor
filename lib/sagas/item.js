import { delay } from 'redux-saga';
import fetch from 'isomorphic-unfetch';
import {
  call,
  put,
  race,
  take,
  takeEvery,
  select,
  all,
} from 'redux-saga/effects';
import { doFetchItems, doSetItems, doFetchItemError } from '../actions/item';
import { API_URL_ITEMS } from '../constants/api';
import {
  ITEMS_POLL_START,
  ITEMS_POLL_STOP,
  ITEMS_FETCH,
} from '../constants/actionTypes';

const ITEM_TYPES = {
  LEGENDARY: 'LEGENDARY',
  STANDARD: 'STANDARD',
  BACKSTAGE_PASS: 'BACKSTAGE_PASS',
  CONJURED: 'CONJURED',
};

/**
 * @typedef Item
 *
 * Item.
 *
 * @type {Object}
 * @property {string} id - Item unique identifier.
 * @property {string} name - Item name.
 * @property {number} sellIn - Sell price.
 * @property {number} quality - Quality (between 0 and 100).
 * @property {string} type - Type of item: "STANDARD", "CONJURED", "BACKSTAGE_PASS" or "LEGENDARY".
 */

/**
 * @typedef ChangingItem
 *
 * Item with trends over time.
 *
 * @type {Item}
 * @property {number} qualityTrend - Difference in quality during the last quality change.
 * @property {number} sellInTrend - Difference in sellIn during the last sellIn change.
 */

/**
 * @typedef ItemSet
 *
 * Items as stored in Redux state: indexed by IDs.
 *
 * @type {Object<string, ChangingItem}
 */

/**
 * Update items based on previous state.
 *
 * TODO Move this logic to the API.
 *
 * @param {ItemSet} prevItems - Previous items indexed by IDs.
 * @param {Item[]} items - New items set from the API.
 * @returns {Array} The new items indexed by IDs (ItemSet), and the number of added, removed and
 * udpated items (number).
 */
export function mergeItems(prevItems, items) {
  return items.reduce(
    ([nextItems, added, removed, updated], item) => {
      const prevItem = prevItems[item.id];
      const qualityTrend = prevItem && prevItem.quality - item.quality;
      const sellInTrend = prevItem && prevItem.sellIn - item.sellIn;
      return [
        // Item set
        {
          ...nextItems,
          [item.id]: {
            ...item,
            qualityTrend:
              qualityTrend || (prevItem ? prevItem.qualityTrend : 0),
            sellInTrend: sellInTrend || (prevItem ? prevItem.sellInTrend : 0),
          },
        },
        // Counts
        prevItem ? added : added + 1,
        prevItem ? removed - 1 : removed,
        qualityTrend || sellInTrend ? updated + 1 : updated,
      ];
    },
    [{}, 0, Object.keys(prevItems).length, 0],
  );
}

/**
 * Fetch items from the API.
 *
 * @returns {Promise<Item[]>} All items.
 */
export async function fetchItems() {
  const debug = process.env.REACT_APP_DEBUG;

  if (debug) {
    console.log(`Fetching items from ${API_URL_ITEMS}`); // eslint-disable-line no-console
  }

  const response = await fetch(API_URL_ITEMS);
  const items = await response.json();

  if (debug) {
    console.log(`Found ${items.length} items.`); // eslint-disable-line no-console
  }

  return items;
}

/**
 * Fetch items from the API and returns the sellable ones only.
 *
 * @returns {Promise<Item[]>} Sellable items.
 */
export async function fetchItemsForSale() {
  return (await fetchItems()).filter(
    item => item.type !== ITEM_TYPES.LEGENDARY,
  );
}

/**
 * Generate two effects:
 * 1. A call to fetchItemsForSale
 * 2. A new ITEMS_SET actions with the items returned by fetchItemsForSale.
 *
 * @returns {Generator} A generator that yields a call to fetchItemsForSale and a ITEMS_SET action.
 */
export function* handleFetchItems() {
  try {
    const items = yield call(fetchItemsForSale);
    const prevItems = yield select(state => state.items);
    const [nextItems] = mergeItems(prevItems, items);
    yield put(doSetItems(nextItems));
  } catch (error) {
    yield put(doFetchItemError(error));
  }
}

/**
 * Generate the following effects infinitely:
 * 1. A new ITEMS_FETCH action.
 * 2. A call to wait() with the given delay.
 *
 * @param {number} ms - Time to wait before next poll, in ms.
 * @returns {Generator} A generator that yields infinitely a ITEMS_SET action and a call to wait.
 */
export function* handlePollItems(ms) {
  while (true) {
    yield put(doFetchItems());
    yield call(delay, ms);
  }
}

/**
 * Generate the following effects infinitely:
 * 1. A wait for the ITEMS_POLL_START action.
 * 2. A race between a handlePollItems call and a wait for ITEMS_POLL_STOP. This means effects from
 *    handlePollItems will be treated infinitely, until a ITEMS_POLL_STOP action is received.
 *
 * @returns {Generator} A generator that yields infinitely a wait forITEMS_POLL_START action and the
 * polling race.
 */
export function* watchPollItemsStart() {
  while (true) {
    const action = yield take(ITEMS_POLL_START);
    yield race([call(handlePollItems, action.delay), take(ITEMS_POLL_STOP)]);
  }
}

/**
 * Create the items poll start/stop saga.
 *
 * @returns {Generator} The items poll start/stop saga.
 */
export default function* itemsSaga() {
  yield all([takeEvery(ITEMS_FETCH, handleFetchItems), watchPollItemsStart()]);
}
