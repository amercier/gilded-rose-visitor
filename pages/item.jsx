import React from 'react';
import { string, number, shape, object } from 'prop-types';
import { connect } from 'react-redux';
import { doFetchItem } from '../lib/actions/item';

/**
 * Item details page.
 *
 * @param {Object} props - React component properties.
 * @param {Item} props.item - Item to display.
 * @param {Item} props.error - Item error, if any.
 * @returns {React.Element} The rendered element.
 */
const Item = ({ item, error }) => (
  <main>
    {item && (
      <>
        <h1>{item.name}</h1>
        <dl>
          <dt>Type</dt>
          <dd>{item.type}</dd>
          <dt>Quality</dt>
          <dd>{item.quality}</dd>
          <dt>Price</dt>
          <dd>{item.sellIn}</dd>
        </dl>
      </>
    )}
    {error && (
      <p>
        An error occured while retrieving your item information. Please retry
        later.
      </p>
    )}
  </main>
);

Item.propTypes = {
  item: shape({
    id: string,
    name: string,
    sellIn: number,
    quality: number,
    type: string,
  }),
  error: object,
};

Item.defaultProps = {
  item: undefined,
  error: undefined,
};

Item.getInitialProps = async ({ ctx }) => {
  const { store, query } = ctx;

  // Start items polling on the browser, or 1 fetch if server-side
  store.dispatch(doFetchItem(query.id));

  return {};
};

/**
 * Map Redux state to <QualityFilter> properties.
 *
 * @param {Object} state - Redux state.
 * @returns {Object} Properties for <QualityFilter> component.
 */
const mapStateToProps = state => ({
  item: state.item,
  error: state.fetchItemError,
});

export { Item };
export default connect(mapStateToProps)(Item);
