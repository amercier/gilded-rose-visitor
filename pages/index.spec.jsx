import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { mount, shallow } from 'enzyme';
import { doStartPollingItems, doFetchItems } from '../lib/actions/item';
import ConnectedIndex, { Index } from '.';

const itemsStub = {
  a: { id: 'a', name: 'ITEM A', quality: 10, sellIn: 10, type: 'STANDARD' },
  b: { id: 'b', name: 'ITEM B', quality: 20, sellIn: 20, type: 'STANDARD' },
  c: { id: 'c', name: 'ITEM C', quality: 30, sellIn: 30, type: 'LEGENDARY' },
};

describe('Index', () => {
  it('renders a list of items', () => {
    const noop = () => {}; // eslint-disable-line require-jsdoc
    const index = shallow(
      <Index
        items={Object.values(itemsStub)}
        cart={[]}
        onAddItem={noop}
        onRemoveItem={noop}
      />,
    );
    expect(index).toMatchSnapshot();
  });
});

describe('Index (Redux-connected)', () => {
  const mockStore = configureStore();

  it('uses values of items object from the store', () => {
    const store = mockStore({
      itemReducer: { items: itemsStub },
      cartReducer: { cart: [] },
    });
    const index = mount(
      <Provider store={store}>
        <ConnectedIndex />
      </Provider>,
    );
    const indexComponent = index.find(Index).first();
    expect(indexComponent.props().items).toEqual(Object.values(itemsStub));
  });

  describe('getInitialProps', () => {
    it('dispatches a ITEMS_FETCH action on the server', () => {
      const items = {};
      const store = mockStore({
        itemReducer: { items },
        cartReducer: { cart: [] },
      });
      Index.getInitialProps({ ctx: { store, isServer: true } });
      expect(store.getActions()).toEqual([doFetchItems(items)]);
    });

    it('dispatches a ITEMS_POLL_START action on the client', () => {
      const items = {};
      const store = mockStore({
        itemReducer: { items },
        cartReducer: { cart: [] },
      });
      Index.getInitialProps({ ctx: { store, isServer: false } });
      expect(store.getActions()).toEqual([doStartPollingItems()]);
    });
  });
});
