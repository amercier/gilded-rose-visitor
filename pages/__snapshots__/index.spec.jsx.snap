// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index renders a list of items 1`] = `
<main>
  <h1>
    Gilded Rose Inn
  </h1>
  <ul>
    <li
      key="a"
    >
      <Link
        as="/i/a"
        href="/item?id=a"
      >
        <a>
          ITEM A
        </a>
      </Link>
      <button
        onClick={[Function]}
        type="button"
      >
        Add to cart
      </button>
    </li>
    <li
      key="b"
    >
      <Link
        as="/i/b"
        href="/item?id=b"
      >
        <a>
          ITEM B
        </a>
      </Link>
      <button
        onClick={[Function]}
        type="button"
      >
        Add to cart
      </button>
    </li>
    <li
      key="c"
    >
      <Link
        as="/i/c"
        href="/item?id=c"
      >
        <a>
          ITEM C
        </a>
      </Link>
      <button
        onClick={[Function]}
        type="button"
      >
        Add to cart
      </button>
    </li>
  </ul>
</main>
`;
